cpp/src/hdbscan/detail/condense.cuh

```diff
@@ -24,11 +24,215 @@
+ *    - Left child inherits parent's label
+ *    - Collapse right subtree, add edges to all its leaf nodes
+ *
+ * we ignore subnodes of the collapsed subtree by using a boolean array
+ */
+template <typename value_idx, typename value_t>
+void _build_condensed_hierarchy(const raft::handle_t& handle,
+                                const value_idx* children,
+                                const value_t* delta,
+                                const value_idx* sizes,
+                                int n_leaves,
+                                int min_cluster_size,
+                                rmm::device_uvector<value_idx>& out_parent,
+                                rmm::device_uvector<value_idx>& out_child,
+                                rmm::device_uvector<value_t>& out_lambda,
+                                rmm::device_uvector<value_idx>& out_size)
+{
+  cudaStream_t stream  = handle.get_stream();
+  value_idx root       = 2 * (n_leaves - 1);
+  value_idx n_samples  = n_leaves;
+  value_idx next_label = n_samples + 1;
+
+  // Copy data to host
+  std::vector<value_idx> h_children(2 * (n_leaves - 1));
+  std::vector<value_t> h_delta(n_leaves - 1);
+  std::vector<value_idx> h_sizes(n_leaves - 1);
+
+  raft::copy(h_children.data(), children, 2 * (n_leaves - 1), stream);
+  raft::copy(h_delta.data(), delta, n_leaves - 1, stream);
+  raft::copy(h_sizes.data(), sizes, n_leaves - 1, stream);
+  handle.sync_stream(stream);
+
+  // host side output arrays
+  std::vector<value_idx> h_out_parent;
+  std::vector<value_idx> h_out_child;
+  std::vector<value_t> h_out_lambda;
+  std::vector<value_idx> h_out_sizes;
+
+  // Get BFS ordering from root
+  std::vector<value_idx> node_list;
+  bfs_from_node(root, n_samples, h_children.data(), node_list);
+
+  std::vector<value_idx> relabel(root + 1, 0);
+  std::vector<bool> ignore(node_list.size(), false);
+  relabel[root] = n_samples;
+
+  // Process nodes in BFS order
+  for (size_t idx = 0; idx < node_list.size(); idx++) {
+    value_idx node = node_list[idx];
+
+    // Skip if already processed or is a leaf
+    if (ignore[node] || node < n_samples) { continue; }
+
+    value_idx left       = h_children[(node - n_samples) * 2];
+    value_idx right      = h_children[(node - n_samples) * 2 + 1];
+    value_t distance     = h_delta[node - n_samples];
+    value_t lambda_value = distance > 0.0 ? 1.0 / distance : std::numeric_limits<value_t>::max();
+
+    value_idx left_count  = left >= n_samples ? h_sizes[left - n_samples] : 1;
+    value_idx right_count = right >= n_samples ? h_sizes[right - n_samples] : 1;
+
+    if (left_count >= min_cluster_size &&
+        right_count >= min_cluster_size) {  // Case 1: Both children are large enough
+      relabel[left] = next_label++;
+      h_out_parent.push_back(relabel[node]);
+      h_out_child.push_back(relabel[left]);
+      h_out_lambda.push_back(lambda_value);
+      h_out_sizes.push_back(left_count);
```

Comment:
Put this logic in a lambda please, it's repeated quite often.

Reply:
done!